Extract jnidispatch native libs

Akirathan opened the pull request:
Closes #13151

### Pull Request Description

- JNA is not dependency of engine anymore - remove it from the `component` directory.
- Extract native libs from `jna-wrapper` in Standard.Microsoft.
  - There is only `jna-wrapper-thin.jar` along with extracted native libs in Std.Microsoft's `polyglot` directory.

### Important Notes
See the newly expected [native_libraries.json](https://github.com/enso-org/enso/pull/13212/files#diff-1eee15dd7c4ac0ce99bfab8f1825aa349a3a00d5c5f64fae47066ea71eb7a7c3)

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [X] The documentation has been updated, if necessary.
- [X] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [X] Unit tests have been written where possible.
- [X] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

Commit:
jna-wrapper/exportedModuleBin extracts native libraries

Commit:
Remove jna native libs from expected native libs from test

Commit:
buildEngineDistribution copies some native libs to the component directory

Commit:
typo in docs

Commit:
std-microsoft depends on the minimized jna-wrapper without native libs

Commit:
jna-wrapper native libs should not be present inside Standard.Microsoft.

JaroslavTulach commented on build.sbt:
```diff
@@ -1416,6 +1416,11 @@ lazy val `language-server-deps-wrapper` = project
+  "Directory where native libraries are extracted to"
+)
+
 lazy val `jna-wrapper` = project
```
I think we should remove JNA from engine/ls. The arguments remain the same as https://github.com/enso-org/enso/issues/7178#issuecomment-1634588128

Akirathan replied:
Doing that on #13223

Akirathan commented:
Blocked by #13223

Commit:
Merge branch 'develop' into wip/akirathan/13151-extract-jnidispatch-native-libs

# Conflicts:
#	test/Base_Tests/data/native_libs.json

Commit:
akka-wrapper does not depend on jna-wrapper

Commit:
file-watcher does not depend on jna-wrapper

Commit:
no jna-wrapper in component directory

Commit:
no jni native libs in component directory

Commit:
extract jni native libs in std-microsoft.

std-microsoft has only jna-wrapper-thin.jar in its polyglot dir, and the extracted native libs

Commit:
fmt

hubertp commented on project/JARUtils.scala:
```diff
@@ -219,4 +219,73 @@ object JARUtils {
+  def readNativeCodeEntriesFromManifest(
+    jarPath: Path
+  ): List[NativeCodeEntry] = {
+    Using(new JarFile(jarPath.toFile)) { jarFile =>
```
it's a bit hard to read this method as it is neither idiomatic Scala nor Java.
I think it would be fairly easy to rewrite it to the former by replacing all those extra returns and nulls with `Option`.

Akirathan replied:
I am sorry but I don't really see an "easy rewrite" to idiomatic Scala code in this case - other than introducing 2 more nestings. IMHO, spaghetti code is less readable than this code. Please, feel free to either provide a suggestion, or push a commit directly so that I know what you consider "idiomatic Scala code" in this case. At this moment, I don't understand reason for "blocking" this PR on this issue. I want to respect your suggestions so before I integrate I am giving you a chance to educate me in this regard.

hubertp replied:
I went medieval on your code, sorry.

hubertp commented on project/JARUtils.scala:
```diff
@@ -219,4 +219,73 @@ object JARUtils {
+  }
+
+  /** @param processor
+    * @param osName
```
?

Akirathan replied:
Docs added in cf0ea254d2af619f00e0c7ebdf91519352579942

hubertp commented on project/StdBits.scala:
```diff
@@ -233,6 +233,86 @@ object StdBits {
+    }
+    val nativeCodeEntries =
+      JARUtils.readNativeCodeEntriesFromManifest(jnaJar.toPath)
+    val (expectedOsName, targetOs) = if (Platform.isWindows) {
```
I feel like I've already seen code that handles OS and architecture naming. Let's keep it in one place, please.

Akirathan replied:
I don't see an easy way how to keep this logic in one place, since every 3rd party dependency uses a different naming conventions. The only way to keep this logic in one place is to refactor the whole file and extract all the similar logic to a separate class. But then again, there will be x different handling for x different 3rd party libraries.

Commit:
Add docs to NativeCodeEntry

hubertp committed:
From Scala with :heart:

hubertp committed:
nit

Akirathan commented:
Failing jobs:
- [JVM Tests (GraalVM CE) (macos, amd64) (Optional)](https://github.com/enso-org/enso/actions/runs/15635806358/job/44063194801?pr=13212#step:14:2712)
  - Failures in FileWatcherTests. Reported in #13280. Unrelated to this PR.
- [standard Library Tests (GraalVM CE) (native) (macos, amd64) (Optional)](https://github.com/enso-org/enso/actions/runs/15635806358/job/44063194860?pr=13212#step:14:1439)
  - Failure in allocating Managed_Resource in Base_Tests. Doesn't seem to be related to this PR as well.

@hubertp Since the aforementioned failures  are most likely not connected to this PR, may I ask you for your approval?
